File: 205-IsomorphicStrings/205-IsomorphicStrings.py
# Last updated: 8/1/2025, 10:38:51 PM
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        mapping = {}  # Mapping from characters in s to characters in t
        
        for char_s, char_t in zip(s, t):
            # If char_s is not in the mapping, add it with char_t
            if char_s not in mapping:
                # Check if char_t is already mapped to another char in s
                if char_t in mapping.values():
                    return False
                mapping[char_s] = char_t
            else:
                # Check if the mapping is consistent
                if mapping[char_s] != char_t:
                    return False
        
        return True

        # st={}
        # ts={}
        # for chars,chart in zip(s,t):
        #      if chars not in st:
        #         st[chars]=chart
        #      elif st[chars]!=chart:
        #         return False
        #      if chart not in ts:
        #         st[chart]=chars
        #      elif st[chart]!=chars:
        #         return False   
        

File: 205-IsomorphicStrings/205-IsomorphicStrings.java
// Last updated: 8/1/2025, 10:38:44 PM
import java.util.HashMap;

class Solution {
    public boolean isIsomorphic(String s, String t) {
        HashMap<Character, Character> map = new HashMap<>();
        if (s.length() != t.length()) return false;
        
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (map.containsKey(ch)) {  
                if (map.get(ch) != t.charAt(i)) return false; 
            } else {
                if (map.containsValue(t.charAt(i))) return false; 
                map.put(ch, t.charAt(i));
            }
        }
        return true;
    }
}
